src/ruby/puppetserver-lib/puppet/server/execution.rb

```diff
@@ -7,31 +7,41 @@ class Puppet::Server::Execution
   def self.initialize_execution_stub
     Puppet::Util::ExecutionStub.set do |command, options, stdin, stdout, stderr|
       if command.is_a?(Array)
+        orig_command_str = command.join(" ")
         binary = command.first
         args = command[1..-1]
       else
+        orig_command_str = command
         binary = command
         args = nil
       end
 
-      # TODO - options is currently ignored - https://tickets.puppetlabs.com/browse/SERVER-74
+      # TODO - options relating to stdout/stderr are not yet handled, see SERVER-74 and
+      #  PUP-6640
 
       # We're going to handle STDIN/STDOUT/STDERR in java, so we don't need
       # them here.  However, Puppet::Util::Execution.execute doesn't close them
       # for us, so we have to do that now.
       [stdin, stdout, stderr].each { |io| io.close rescue nil }
 
-      execute(binary, args)
+      execute(orig_command_str, binary, options, args)
     end
   end
 
-  def self.execute(command, args = nil)
+  private
+  def self.execute(orig_command_str, command, options, args)
     if args && !args.empty?
       result = ShellUtils.executeCommand(command, args.to_java(:string))
     else
       result = ShellUtils.executeCommand(command)
     end
 
-    Puppet::Util::Execution::ProcessOutput.new(result.getOutput, result.getExitCode)
+    # TODO - not all options from Puppet::Util::Execution are supported yet, see SERVER-74
+
+    if options[:failonfail] and result.exit_code != 0
```

Comment:
I copied this straight out of the puppet code, after discussing with @joshcooper about whether there was a sane way to avoid the duplication.  (There wasn't.)  I'd like to keep it as similar to the puppet code as possible so that it's as easy as possible to tell if something changes upstream that we need to port over.  So, I think the best choices here are:

1. Add a comment explaining that
2. Change it here, and in the puppet code as well (seems like overkill, but I'd be happy to do it)
3. Leave it as-is.